driver/src/test/java/org/neo4j/driver/v1/EncryptionTest.java

```diff
@@ -0,0 +1,192 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.driver.v1;
+
+import io.netty.bootstrap.ServerBootstrap;
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.ssl.SslHandler;
+import org.junit.jupiter.api.Test;
+
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.neo4j.driver.v1.exceptions.ServiceUnavailableException;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.startsWith;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.junit.MatcherAssert.assertThat;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.neo4j.driver.internal.async.BoltProtocolUtil.BOLT_MAGIC_PREAMBLE;
+
+class EncryptionTest
+{
+    @Test
+    void shouldOperateWithNoEncryptionWhenItIsOptionalInTheDatabase() throws Throwable
+    {
+        testDriverEncryption( false );
+    }
+
+    @Test
+    void shouldOperateWithEncryptionWhenItIsOptionalInTheDatabase() throws Throwable
+    {
+        testDriverEncryption( true );
+    }
+
+    private void testDriverEncryption( boolean isEncrypted ) throws Throwable
+    {
+        DummyNettyServer server = new DummyNettyServer();
+        int port = server.start();
+
+        Config config = newConfig( isEncrypted );
+        RuntimeException e = assertThrows( RuntimeException.class, () -> GraphDatabase.driver( "bolt://localhost:" + port, config ).close() );
+        server.stop();
+
+        server.assertClientEncryption( isEncrypted );
+        assertThat( e, instanceOf( ServiceUnavailableException.class ) );
+        assertThat( e.getMessage(), startsWith( "Connection to the database terminated" ) );
+    }
+
+    private static Config newConfig( boolean withEncryption )
+    {
+        return withEncryption ? configWithEncryption() : configWithoutEncryption();
+    }
+
+    private static Config configWithEncryption()
+    {
+        return Config.build().withEncryption().toConfig();
+    }
+
+    private static Config configWithoutEncryption()
+    {
+        return Config.build().withoutEncryption().toConfig();
+    }
+
+    enum ClientSocketStatus
+    {
+        Encrypted,
+        Unencrypted,
+        Unknown,
+        FailedToRecognize
+    }
+
+    public final class DummyNettyServer
+    {
+        private EventLoopGroup bossGroup;
```

Comment:
Single event loop group can be used for the whole dummy server. There exists `ServerBootstrap#group(EventLoopGroup)` method and we most likely do not need separate groups in a test.